### ISSUE TITLE
Replace TypeVar annotations in Generic with concrete type

### ISSUE THREAD

**Issue by spflueger** [python/typing#762]
Hi,

first of all thanks for the great `typing` package!

I'm using a generic class in the following way

```python3

T = TypeVar("T")

class MyGeneric(Generic[T]):
    def __call__(
        self,
        ingoing_edges: List[T],
        outgoing_edges: List[T],
    ) -> bool:
        return sum(ingoing_edges) == sum(outgoing_edges)

IntGeneric = MyGeneric[int]()

IntGeneric([1,2,3], [3,3])
```

This all works fine. However, I'm expecting the type annotations for the `IntGeneric.__call__` to be `List[int]`, but they are `List[~T]`. Would replacing the generic type annotations with the "concrete" type follow the idea/concept of typing, or is there some reason I'm not thinking of that intends this behavior?

**Comment by gvanrossum** [comment]
That's just not how it works. Type annotations are statically created when the module is imported, and calling `MyGeneric[int]` is supposed to be a very lightweight operation -- it can't go around and create a new class object with all new method objects whose `__annotations__` is updated.  I suppose you could create a helper function that does the substitutions, but the question remains, why? You say "I'm expecting the type annotations [...] to be `List[int]`" -- but what difference does it make to you? What important use case would be enabled by that functionality?

I don't think this is a desirable feature.

**Comment by spflueger** [comment]
Oh okay, I get it. Thanks for the clarification.

> I suppose you could create a helper function that does the substitutions, but the question remains, why? You say "I'm expecting the type annotations [...] to be List[int]" -- but what difference does it make to you? What important use case would be enabled by that functionality?

Probably not an important use case. I'm experimenting with rule based system of mine and using the type annotations of rules (callables) to determine their conditions for execution. A data structure contains all different data types. Some other code dynamically determines from the type annotations, which data types can be fed into the various rules.
Then I noticed that a Generic would fit nicely for some subset of the rules, and that's when I got stuck at this point. So I guess I was just naively expecting the type to be replaced ^^. But I'm to achieve this with a parameterized decorator.
